### PR TITLE
Improve parallelism of repartition operator

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Close https://github.com/apache/arrow-datafusion/issues/6290

# Rationale for this change

I was testing query performance for https://github.com/apache/arrow-datafusion/issues/6278 and noticed that only. a single core was being used on a query entirely in memory. When I spent some time looking into it, the plan looked correct with repartitioning but for some reason it wasn't properly repartitoning 

# What changes are included in this PR?

Don't yield on *every* batch -- yield only after we have made some decent progress (in this case at least `partition_count` batches)

# Are these changes tested?
I manually tested this -- I will add a benchmark for it shortly

My manual test results are:

On main (Keeps only 1 core busy for most of the time)
```
4 rows in set. Query took 10.631 seconds.
```

With this PR (keeps the cores much busier)
```
4 rows in set. Query took 3.705 seconds.
```

# Are there any user-facing changes?

### CODE REVIEW

**Comment by alamb** [comment]
cc @crepererum

File: datafusion/core/src/physical_plan/repartition/mod.rs [hunk 1]
```diff
@@ -532,9 +541,28 @@ impl RepartitionExec {
                 timer.done();
             }
 
-            // If the input stream is endless, we may spin forever and never yield back to tokio. Hence let us yield.
-            // See https://github.com/apache/arrow-datafusion/issues/5278.
-            tokio::task::yield_now().await;
+            // If the input stream is endless, we may spin forever and
```
Comment (alamb) [inline datafusion/core/src/physical_plan/repartition/mod.rs:544]: I would love some thoughts from reviewers about better heuristics here -- as the comments say I am happy with this heuristic for round robin partitioning but there may be a better way when hash partitioning (like ensure that all channels have at least one batch 🤔 )

Comment (tustvold) [reply datafusion/core/src/physical_plan/repartition/mod.rs:544]: You could use https://docs.rs/tokio/latest/tokio/task/fn.consume_budget.html but I'm honestly a little confused by this. "we may spin forever" would imply an issue with unbounded receivers, not an issue with the repartition operator?

TLDR I'd vote to not yield at all, I don't agree that this fixes #5278 rather just papers over it with a dubious fix

Comment (alamb) [reply datafusion/core/src/physical_plan/repartition/mod.rs:544]: I think if the tokio executor has only a single thread and the input stream can provide data infinitely, without a yield it will buffer the entire input which seems non ideal 

I agree https://github.com/apache/arrow-datafusion/issues/5278 as described seems somewhat more like "when we used blocking IO with a single tokio thread it blocked everything" -- as described on https://github.com/apache/arrow-datafusion/issues/5278#issuecomment-1432901387

Comment (tustvold) [reply datafusion/core/src/physical_plan/repartition/mod.rs:544]: > it will buffer the entire input which seems non ideal

That seems like a bug in whatever is using unbounded buffers, which I thought we had removed the last of?

Comment (crepererum) [reply datafusion/core/src/physical_plan/repartition/mod.rs:544]: You can call it a bug or a design issue of DF / tokio. But if you run two spawned tasks and one never returns to Tokio then the other will never run. Unbounded buffers are NOT avoidable in the current DF design, because you cannot predict tokio scheduling and hash outputs. So the fix here is adequate. `consume_budget` would be the better solution but it's an unstable tokio feature, so that's not usable.

**Comment by alamb** [comment]
FYI @metesynnada I don't know if you have any thoughts on this approach (or the yield in general, as you reported https://github.com/apache/arrow-datafusion/issues/5278 initially)

**Comment by ozankabak** [comment]
Hi @alamb, @metesynnada is unavailable for this month, he will be back at the end of the month. I checked this PR out and it  looks better than the status quo to me.

**Comment by alamb** [comment]
I plan to leave this open for another day or two for comments and then will merge in

**Comment by alamb** [comment]
I merged up from main -- and once CI passes I plan to merge this PR

**Comment by crepererum** [comment]
Side note: when I've added the yield statement I was wondering if this would be too much overhead, but my assumption was the the batches would hopefully be big enough so that wouldn't matter. Seems that I was wrong 😅

**Comment by alamb** [comment]
>  Side note: when I've added the yield statement I was wondering if this would be too much overhead, but my assumption was the the batches would hopefully be big enough so that wouldn't matter. Seems that I was wrong 😅

I think the fact that the source in this case is a MemoryExec (where all the data is already in memory and can be provided almost instantaneously hurts us)
